pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+
+Consider the following sequence of examples. In the left, a JSON Schema with a
+single nested schema resource. In the right, the corresponding directed graph
+representations for the root schema resource called
+`https://example.com/person` and the nested schema resource called
+`https://example.com/surname`. At each step of the evaluation process, we gray
+out the parts of the schema and of the directed graph that are not part of the
+lexical scope.
+
+The evaluation process starts with the top level schema. The lexical scope at
+that point is the root schema resource, and the nested schema resource is out
+of scope. 
+
+![The lexical scope of a JSON Schema (1)](/img/posts/2024/understanding-lexical-dynamic-scopes/lexical-scope-1.webp)
+
+Then, we enter the `properties` applicator, and if the instance defines a
+`firstName` property, we get into the subschema at `/properties/firstName`.
+This subschema is part of the root schema resource (as it does not declare its
+own identifier), so the lexical scope remains the same as the previous step. 
+
+![The lexical scope of a JSON Schema (2)](/img/posts/2024/understanding-lexical-dynamic-scopes/lexical-scope-2.webp)
+
+Finally, if the instance defines a `lastName` property, we follow the
+`properties` applicator into the subschema at `/properties/lastName`. This
+subschema defines a new schema resource, so the lexical scope at this point is
+the nested schema resource, and the root schema resource is out of scope.
+
+![The lexical scope of a JSON Schema (3)](/img/posts/2024/understanding-lexical-dynamic-scopes/lexical-scope-3.webp)
+
+Note that by definition, the lexical scope of any subschema can be statically
+determined without taking
+[instances](https://json-schema.org/learn/glossary#instance) into account, just
+as we did here.
+
+### Lexical Scope and Anchors
+
+As another practical example, consider the `$anchor` keyword that defines a
+[location-independent
+identifier](https://json-schema.org/draft/2020-12/json-schema-core#section-8.2.2)
+for a schema. This keyword not only affects the schema object it is defined in,
+but its lexical scope. This is why declaring the same anchor identifier more
+than once in the same schema resource is an error (a clash in the lexical
+scope), while it is possible to declare the same anchor identifier on different
+schema resources (as the lexical scopes are different):
+
+![Example of anchors within and across lexical scopes](/img/posts/2024/understanding-lexical-dynamic-scopes/scopes-static-anchors.webp)
+
+### Following References
+
+When the evaluation process encounters a reference keyword, it *abandons* the
+lexical scope of the reference schema and enters the lexical scope of the
+*destination* schema.
+
+If the reference points to a subschema *within* the same schema resource, the
+lexical scope effectively remains the same. Coming back to the graph analogy,
+the evaluation process remains at the same node. However, if the reference
```

Comment:
> Coming back to the graph analogy, the evaluation process remains at the same node.

This isn't right.  The evaluation process always changes _nodes_ by following the `$ref`.  It's just that in this case, the lexical scope remains the same.

Your image is correct: the lexical scope remains the same, but the evaluated node changes from `/items` to `/$defs/person-name`.

Reply:
We are considering nodes to be schema resources. If we take into account references within a single schema resource, the reference changes subschemas, but not schema resources, so the node (at least how we define) remains the same, right?

Reply:
Oh, I see: "node" refers to the graph here.  That wasn't clear to me at first.  I was thinking "JSON node".  I think your rewording makes that clearer. 👍

Reply:
Great! Let me clarify it a bit more on the post anyway